### ISSUE TITLE
Skipping generation step of compilation if the code was changed

### ISSUE THREAD

**Issue by RealyUniqueName** [HaxeFoundation/haxe#8209]
Following  the discussion in f0be760d8673a3bbb621bf6da956c0950c0a7b9a

I think it's not a compiler's responsibility to skip something when you actually want to avoid the whole compilation.
It looks like a compiler is solving a 3rd-party project workflow issue.

**Comment by ncannasse** [comment]
For me the argument does not stand, that's already what C++ compiler does when they skip linking when no object file has been modified.

**Comment by ncannasse** [comment]
But I agree we should:
- add a print when running the server in verbose mode
- allow to disable the behavior with a define (just in case)
but IMHO it should still be the default.

**Comment by Gama11** [comment]
`onAfterGenerate` (which came in the other discussion) is probably not an issue here - it's already being called even when no generation is done with `--no-output`, so the same should simply be done here too (probably already is the case).

https://github.com/HaxeFoundation/haxe/blob/93b7c76f079526ab85e1d35e77652f5913c3404d/std/haxe/macro/Context.hx#L335

**Comment by nadako** [comment]
In theory I agree that it's not technically a compiler concern, but I think it's simply pragmatic and useful in real world. Also, the cache server knows about all the modules and their invalidations and it would be annoying to re-implement the behaviour in a third-party tool. One option would be to provide a compiler API to get the info about cache, but at the moment it feels like over-engineering.

I also agree with the @ncannasse's points about verbose mode note and disabling define flag.

**Comment by Gama11** [comment]
Something I didn't expect at all, but makes a lot of sense: vshaxe compilation actually seems to benefit **immensely** from this change.

My usual workflow is to simply press F5 to build everything through the server, and that triggers builds for multiple sub-projects:

- the vshaxe VSCode extension itself
-  haxe-language-server
- eval-debugger

With this change, compilation has turned from "waiting a few seconds" to "basically instant" when there weren't any changes. Even if there were changes, it makes a lot of sense to skip the generation for the other projects that weren't changed.

**Comment by ncannasse** [comment]
Yes that's the whole point of the feature ;)

**Comment by Gama11** [comment]
I guess, but I don't think the multi-project-compilation use case was mentioned before? :)

**Comment by Gama11** [comment]
Hm, it seems there's project setups that are broken by this behavior change because they rely on a `-cmd` being called. For instance https://github.com/Gama11/HaxeRepro uses `-cmd` to run the program after compilation, and that should work even if no source files changed.

Something like adding a `-D dump-dependencies` flag also no longer works without touching some source files.

**Comment by Gama11** [comment]
This even seems to break any project that uses vshaxe's completion cache, because `--no-output` compilations seem to count. So you can't even build a newly loaded project without changing something in the sources.

Given how many edge cases / issues there seem to be, I have reverted the change for now.

**Comment by ncannasse** [comment]
I'm listing the fixes require:
- add a flag to disable the behavior
- add a print in verbose mode to tell it's not skipped
- always dump even if there's no change
- always run cmd even if no change : this one is questionable, as some cmd might actually move the output file elsewhere, I would instead not call commands and if you want to keep current behavior you would use the flag
- compare the context timestamp vs the current output file to know if the output needs to be rebuild (will fix --no-output)
Anything else ?

**Comment by Gama11** [comment]
>always dump even if there's no change

In theory, it could probably be detected whether a dump needs to happen. Not sure that's worth the complexity given that it's mainly a debugging feature though.

>always run cmd even if no change : this one is questionable [...]

Yeah, it's not clear what the behavior here should be, both use cases exist and are valid. :/

>compare the context timestamp vs the current output file to know if the output needs to be rebuild 

That seems like a strange approach, and likely doesn't work for targets that don't output a single file? I think gencpp specifically avoids changing timestamps of unchanged files.

Couldn't `needs_generation` just not be set to `false` if there's `--no-output`?

**Comment by ncannasse** [comment]
> > compare the context timestamp vs the current output file to know if the output needs to be rebuild
> 
> That seems like a strange approach, and likely doesn't work for targets that don't output a single file? I think gencpp specifically avoids changing timestamps of unchanged files.

Right, we will find another way to make sure that we mark the context has already generated

**Comment by ncannasse** [comment]
Given this significantly improves user experience, I would like to push that for 4.0

**Comment by Simn** [comment]
I'm not opposed to this in principle but worry about how exactly to implement it. Note that `--no-output` currently implies not running `--interp`, which is consistent but gets in the way here.

Your approach to only generate something if we parsed a file is too brittle either way. Consider something like this:

```haxe
class Main {
	static function main() {
		trace(generateTimestamp());
	}

	macro static function generateTimestamp() {
		return macro $v{Sys.time()};
	}
}
```

Dumb example of course, but it shows that output can change even if we never parse any file.

**Comment by ncannasse** [comment]
@Simn we're in the compiler cache context. In this case if Main is not changed the timestamp will stay the same. I'm not much talking about code compilation here, it's mostly about skipping code generation when no change has been made in compilation since last (already done) generation.

**Comment by Simn** [comment]
Right, the point I wanted to make (but didn't make with that example) is that we're equating "the compiler parsed something" with "the output changed". At the very least we should check that we actually typed something, i.e. the `type_module_hook` function in server.ml returned `None`.

**Comment by Simn** [comment]
Given that this isn't a bugfix I'm moving it to 4.1.

**Comment by ncannasse** [comment]
I've submitted an HL specific implementation for this.

**Comment by RealyUniqueName** [comment]
I'm now not opposed to this change, but reducing reliability worries me because of this:

> always run cmd even if no change : this one is questionable, as some cmd might actually move the output file elsewhere, I would instead not call commands and if you want to keep current behavior you would use the flag

I see this as a source of confusion for Haxe users because the compiler can't predict the intention of `--cmd` commands. And those commands could rely on the output being generated every time.

We can make the proposed behavior an opt-in so users would have to explicitly state they are aware of the behavior change.
